capsules/src/button.rs

```diff
@@ -151,14 +167,19 @@ impl<'a, G: hil::gpio::Pin + hil::gpio::PinCtl> Driver for Button<'a, G> {
 impl<'a, G: hil::gpio::Pin> Client for Button<'a, G> {
     fn fired(&self, pin_num: usize) {
         // read the value of the pin
-        let pins = self.pins.as_ref();
-        let pin_state = pins[pin_num].read();
+        let pin_state = self.pins[pin_num].read();
+        let interrupt_count = Cell::new(0);
 
         // schedule callback with the pin number and value
         self.callback.each(|cntr| {
             cntr.0.map(|mut callback| if cntr.1 & (1 << pin_num) != 0 {
+                interrupt_count.set(interrupt_count.get() + 1);
                 callback.schedule(pin_num, pin_state as usize, 0);
             });
         });
+
+        if interrupt_count.get() == 0 {
+            self.pins[pin_num].disable_interrupt();
+        }
```

Comment:
it's a bit of a hack... a process that was previously subscribed might have died... so iterating when a interrupt happens let's us catch that lazily